packages/common/infra/src/blocksuite/migration/fixing.ts

```diff
@@ -12,36 +12,41 @@ import { applyUpdate, encodeStateAsUpdate } from 'yjs';
 // - when fetching the rows of this doc using the doc id === page id,
 //   it will return empty since there is no updates associated with the page id
 export function guidCompatibilityFix(rootDoc: YDoc) {
-  let changed = false;
-  transact(rootDoc, () => {
-    const meta = rootDoc.getMap('meta') as YMap<unknown>;
-    const pages = meta.get('pages') as YArray<YMap<unknown>>;
-    pages?.forEach(page => {
-      const pageId = page.get('id') as string | undefined;
-      if (pageId?.includes(':')) {
-        // remove the prefix "space:" from page id
-        page.set('id', pageId.split(':').at(-1));
-      }
-    });
-    const spaces = rootDoc.getMap('spaces') as YMap<YDoc>;
-    spaces?.forEach((doc: YDoc, pageId: string) => {
-      if (pageId.includes(':')) {
-        const newPageId = pageId.split(':').at(-1) ?? pageId;
-        const newDoc = new YDoc();
-        // clone the original doc. yjs is not happy to use the same doc instance
-        applyUpdate(newDoc, encodeStateAsUpdate(doc));
-        newDoc.guid = doc.guid;
-        spaces.set(newPageId, newDoc);
-        // should remove the old doc, otherwise we will do it again in the next run
-        spaces.delete(pageId);
-        changed = true;
-        console.debug(
-          `fixed space id ${pageId} -> ${newPageId}, doc id: ${doc.guid}`
-        );
-      }
+  function doFix() {
+    transact(rootDoc, () => {
+      const meta = rootDoc.getMap('meta') as YMap<unknown>;
+      const pages = meta.get('pages') as YArray<YMap<unknown>>;
+      pages?.forEach(page => {
+        const pageId = page.get('id') as string | undefined;
+        if (pageId?.includes(':')) {
+          // remove the prefix "space:" from page id
+          page.set('id', pageId.split(':').at(-1));
+        }
+      });
+      const spaces = rootDoc.getMap('spaces') as YMap<YDoc>;
+      spaces?.forEach((doc: YDoc, pageId: string) => {
+        if (pageId.includes(':')) {
+          const newPageId = pageId.split(':').at(-1) ?? pageId;
+          const newDoc = new YDoc();
+          // clone the original doc. yjs is not happy to use the same doc instance
+          applyUpdate(newDoc, encodeStateAsUpdate(doc));
+          newDoc.guid = doc.guid;
+          spaces.set(newPageId, newDoc);
+          // should remove the old doc, otherwise we will do it again in the next run
+          spaces.delete(pageId);
+          console.debug(
+            `fixed space id ${pageId} -> ${newPageId}, doc id: ${doc.guid}`
+          );
+        }
+      });
     });
-  });
-  return changed;
+  }
+
+  doFix();
+
+  // do fix every time when meta or spaces changed
+  rootDoc.getMap('meta').observe(() => doFix());
+  rootDoc.getMap('spaces').observe(() => doFix());
```

Comment:
need to unobserve if you change the function (and fixWorkspaceVersion).
also rename the function to make sure it is not being misused?

Reply:
Because we might be sync incorrect data from other nodes, we run the fix repeatedly with each update.

We don't need `unobserve` because we run the fix repeatedly.

Reply:
I thought we may need to unobserve these on switching or deleting a workspace, but after reading the related code it seems the assumption is that a workspace will be opened once only.
I think it is still a good practice that return disposable for any event binding functions though to make sure this function is not being misused.